### PR TITLE
Import Assertions implementation

### LINKED ISSUE #223: Implement "Import Assertions" as like in esprima-next (when Spec reaches Stage 4)
Copy implementation when Spec reaches Stage 4

Implementation:
https://github.com/node-projects/esprima-next/pull/21/commits/a82b4abe61e48586fb39af7db4f4a72c4f800f4d

Spec in EsTree:
https://github.com/estree/estree/blob/master/stage3/import-assertions.md

### PR DESCRIPTION
**Pull request by Xicy**
1. Import assertions implemented according to #223.
2. Updated **Test262** test cases with latest sha _[53d6cd6d463df461e1c506e0d2be4e36de0ef6fa](https://github.com/tc39/test262/tree/53d6cd6d463df461e1c506e0d2be4e36de0ef6fa)_
3. Added fixture tests from [babel tests](https://github.com/babel/babel/tree/main/packages/babel-parser/test/fixtures/experimental/import-assertions) (output tree regenerated via manual control)

fixes #223

### CODE REVIEW

File: src/Esprima/Ast/ExportAllDeclaration.cs [hunk 1]
```diff
@@ -10,22 +10,39 @@ public sealed class ExportAllDeclaration : ExportDeclaration
...
         protected internal override Node Accept(AstVisitor visitor)
         {
             return visitor.VisitExportAllDeclaration(this);
         }
+
+        private IEnumerable<Node> CreateChildNodes()
+        {
+            yield return Source;
+            
+            if(Exported is not null){
```
Comment (sebastienros) [inline src/Esprima/Ast/ExportAllDeclaration.cs:38]:
```suggestion
            if(Exported is not null)
            {
```

File: src/Esprima/JavascriptParser.cs [hunk 2]
```diff
@@ -4974,20 +5027,21 @@ private ExportDeclaration ParseExportDeclaration()
                         declaration = ThrowUnexpectedToken<StatementListItem>(_lookahead);
                         break;
                 }
-
-                exportDeclaration = Finalize(node, new ExportNamedDeclaration(declaration, new NodeList<ExportSpecifier>(), null));
+
```
Comment (sebastienros) [inline src/Esprima/JavascriptParser.cs:5030]:
```suggestion

```

File: src/Esprima/Utils/AstJson.cs [hunk 3]
```diff
@@ -770,6 +773,9 @@ protected internal override ChainExpression VisitChainExpression(ChainExpression
                 Member("declaration", exportNamedDeclaration.Declaration);
                 Member("specifiers", exportNamedDeclaration.Specifiers);
                 Member("source", exportNamedDeclaration.Source);
+                if(exportNamedDeclaration.Assertions.Count > 0) {
```
Comment (sebastienros) [inline src/Esprima/Utils/AstJson.cs:776]:
```suggestion
                if (exportNamedDeclaration.Assertions.Count > 0)
                {
```

File: src/Esprima/Utils/AstJson.cs [hunk 4]
```diff
@@ -757,6 +757,9 @@ protected internal override ChainExpression VisitChainExpression(ChainExpression
             {
                 Member("source", exportAllDeclaration.Source);
                 Member("exported", exportAllDeclaration.Exported);
+                if(exportAllDeclaration.Assertions.Count > 0) {
```
Comment (sebastienros) [inline src/Esprima/Utils/AstJson.cs:760]:
```suggestion
                if (exportAllDeclaration.Assertions.Count > 0)
                {
```

File: src/Esprima/Utils/AstJson.cs [hunk 5]
```diff
@@ -790,17 +796,34 @@ protected internal override Import VisitImport(Import import)
         {
             using (StartNodeObject(import))
             {
+                if(import.Attributes is not null) {
```
Comment (sebastienros) [inline src/Esprima/Utils/AstJson.cs:799]:
```suggestion
                if (import.Attributes is not null)
                {
```

File: src/Esprima/Utils/AstJson.cs [hunk 6]
```diff
@@ -790,17 +796,34 @@ protected internal override Import VisitImport(Import import)
...
+            return importAttribute;
+        }
 
         protected internal override ImportDeclaration VisitImportDeclaration(ImportDeclaration importDeclaration)
         {
             using (StartNodeObject(importDeclaration))
             {
                 Member("specifiers", importDeclaration.Specifiers, e => (Node) e);
                 Member("source", importDeclaration.Source);
+                if(importDeclaration.Assertions.Count > 0) {
```
Comment (sebastienros) [inline src/Esprima/Utils/AstJson.cs:824]:
```suggestion
                if (importDeclaration.Assertions.Count > 0)
                {
```

**Comment by jogibear9988** [comment]
love this !

**Comment by jogibear9988** [comment]
Have you implemented it excatly same way I did, or did you change a few things? (Maybe you finde some issues in my esprima-next implementation?) If so let me know so I could fix

File: test/Esprima.Tests.Test262/allow-list.txt [hunk 7]
```diff
@@ -2322,15 +2382,9 @@ test/language/module-code/export-expname-from-unpaired-surrogate.js(strict mode)
 test/language/module-code/export-expname-import-unpaired-surrogate.js(strict mode)
 test/language/module-code/export-expname-string-binding.js(strict mode)
 test/language/module-code/export-expname-unpaired-surrogate.js(strict mode)
-test/language/module-code/import-assertion-empty.js(strict mode)
 test/language/module-code/import-assertion-key-identifiername.js(strict mode)
```
Comment (lahma) [inline test/Esprima.Tests.Test262/allow-list.txt:2385]: so still some cases that don't work?

**Review by lahma — COMMENTED** [review]
Great initiative! Can you please rebase and check Sebastien's comments?

**Comment by Xicy** [comment]
> Have you implemented it excatly same way I did, or did you change a few things? (Maybe you finde some issues in my esprima-next implementation?) If so let me know so I could fix

Exactly same version of what you implemented. I will investigate 3 tests that mention by @lahma, If I find some improvement, I will inform you.

>test/language/module-code/import-assertion-key-identifiername.js(strict mode)
>test/language/module-code/import-assertion-key-string-double.js(strict mode)
>test/language/module-code/import-assertion-key-string-single.js(strict mode)

File: src/Esprima/JavascriptParser.cs [hunk 8]
```diff
@@ -4694,6 +4702,67 @@ private Literal ParseModuleSpecifier()
...
+            var attributes = new ArrayList<ImportAttribute>();
+            if (_lookahead.Value is not ("assert"))
+            {
+                return attributes;
+            }
+
+            NextToken();
+            Expect("{");
+
+            var paramMap = new Dictionary<string?, bool>();
```
Comment (lahma) [inline src/Esprima/JavascriptParser.cs:4716]: HashSet could suffice?

File: src/Esprima/JavascriptParser.cs [hunk 9]
```diff
@@ -4694,6 +4702,67 @@ private Literal ParseModuleSpecifier()
...
+                {
+                    case Identifier identifier:
+                        key = identifier.Name;
+                        break;
+                    case Literal literal:
+                        key = literal.StringValue;
+                        break;
+                }
+
+                if (paramMap.ContainsKey(key))
```
Comment (lahma) [inline src/Esprima/JavascriptParser.cs:4732]: With HashSet could check if Add returns false, then it's duplicate

**Review by lahma — APPROVED** [review]
Great work!

**Pull request merged**